lib/nats/io/client.rb

```diff
@@ -1407,39 +1420,12 @@ def process_connect_init
 
       case
       when (server_using_secure_connection? and client_using_secure_connection?)
-        tls_context = nil
-
-        if @tls
-          # Allow prepared context and customizations via :tls opts
-          tls_context = @tls[:context] if @tls[:context]
-        else
-          # Defaults
-          tls_context = OpenSSL::SSL::SSLContext.new
-
-          # Use the default verification options from Ruby:
-          # https://github.com/ruby/ruby/blob/96db72ce38b27799dd8e80ca00696e41234db6ba/ext/openssl/lib/openssl/ssl.rb#L19-L29
-          #
-          # Insecure TLS versions not supported already:
-          # https://github.com/ruby/openssl/commit/3e5a009966bd7f806f7180d82cf830a04be28986
-          #
-          tls_context.set_params
-        end
-
-        # Setup TLS connection by rewrapping the socket
-        tls_socket = OpenSSL::SSL::SSLSocket.new(@io.socket, tls_context)
-
-        # Close TCP socket after closing TLS socket as well.
-        tls_socket.sync_close = true
-
-        # Required to enable hostname verification if Ruby runtime supports it (>= 2.4):
-        # https://github.com/ruby/openssl/commit/028e495734e9e6aa5dba1a2e130b08f66cf31a21
-        tls_socket.hostname = @hostname
-
-        tls_socket.connect
-        @io.socket = tls_socket
+        @io.setup_tls!
       when (server_using_secure_connection? and !client_using_secure_connection?)
         raise NATS::IO::ConnectError.new('TLS/SSL required by server')
-      when (client_using_secure_connection? and !server_using_secure_connection?)
+      # Server requiring TLS/SSL over websocket but not requiring it over standard protocol
+      # doesn't send `tls_required` in its INFO so we need to check the URI scheme.
+      when (client_using_secure_connection? and !server_using_secure_connection? and @uri.scheme != "wss")
```

Comment:
This is a little bit surprising behavior of the NATS server and e.g. demo.nats.io behaves like this: it doesn't send `require_tls: true` in its `INFO` over secure websockets even if it seems not to be configured to accept non-secure websocket connections at all.